IPIP/0000-double-hash-dht.md

```diff
@@ -0,0 +1,333 @@
+- **`HASH2`** is defined as `SHA256(bytes("CR_DOUBLEHASH") || MH)`. It represents the location of the Kademlia keyspace for the Provider Record associated with `CID`. `HASH2` is represented as a 32-byte array. `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)`.
+- **Content Provider** is the node storing some content, and advertising it to the DHT.
+- **DHT Servers** are nodes running the IPFS public DHT. In this documents, DHT Servers mostly refer to the DHT Servers storing the Provider Records associated with specific `CID`s, and not the DHT Servers helping routing lookup requests to the right keyspace location. 
+- **Client** is an IPFS client looking up a content identified by an already known `CID`.
+- **Publish Process** is the process of the Content Provider communicating to the DHT Servers that it provides some content identified by `CID`.
+- **Lookup Process** is the process of the Client retrieving the content identified by `CID`.
+- **`PeerID`** s define stable [peer identities](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md). The `PeerID` is derived from the node's cryptographic public key.
+- **`multiaddrs`** are the [network addresses](https://github.com/libp2p/specs/tree/master/addressing) associated with a `PeerID`. It represents the location(s) of the peer.
+- **`KeyPrefix`** is defined as a prefix of length `l` bits of `HASH2`. `KeyPrefix` is represented by a `byte` concatenated with a variable sized array of bytes, containing at most 32 bytes. The leading `byte` represents the binary representation of `l - 1`, making prefixes of length `256` possible, but not prefixes of length `0`. The trailing byte array is of length `ceil(l/8)` bytes, and its content is the bits prefix right padded with zeros.
+- **`ServerKey`** is defined as `SHA256(bytes("CR_SERVERKEY") || MH)`. It is derived from the `MH`. The Content Provider communicates `ServerKey` to the DHT Servers during the Publish Process. The DHT Servers use it to encrypt the data sent to the Client during the lookup process. `ServerKey` is represented as a 32-byte array.
+- **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.
+- **`CPPeerID`** is the `PeerID` of the Content Provider for a specific `CID`.
+- **`EncPeerID`** is the result of the encryption of `CPPeerID` using `MH` as encryption key and a random nonce `AESGCM(MH, Nonce, CPPeerID)`. `EncPeerID` contains the [varint](https://github.com/multiformats/multicodec/blob/master/table.csv#L69) of the encryption algorithm used (AES), the bytes array of the encrypted payload, and the `Nonce`. `Nonce` is a randomly generated 12-byte array. The format of `EncPeerID` is [`0xa2`, `Nonce`, `AESGCM(MH, Nonce, CPPeerID)`].
```

Comment:
The varint `0xa2`, I believe, refers to 256-bit AES symmetric key. It seems odd using it here, because what it is being prepended to is _not_ 256-bit AES symmetric key. It is an AESGSM encrypted value.

I recommend defining a new varint for such records. That would then make it straightforward to interpret that the next 12 bytes after the varint is nonce.

Reply:
Created a PR to add a specific `aes-gcm-256` varint: https://github.com/multiformats/multicodec/pull/314